## Part 3 - Merging

Let's import the cleaned CollegeData and cleaned USNews rankings dataframes:

In [1]:
import numpy as np
import pandas as pd
from fuzzywuzzy import process
from collegedata_names import school_rename_dict

COLLEGEDATA_CLEAN_CSV_PATH = 'data/collegedata_clean.csv'
USNEWS_CLEAN_CSV_PATH = 'data/usnews_clean.csv'

cols = ['Name', 'State', 'City']

df = pd.read_csv(COLLEGEDATA_CLEAN_CSV_PATH, index_col = cols)
rank_df = pd.read_csv(USNEWS_CLEAN_CSV_PATH, index_col = cols)

len(rank_df)

1381

Since we have less ranks than we have schools, let's try simply left joining the usnews_df to collegedata_df on the common Name, State, and City columns to create a new dataframe 'df'. Before we do that, let's see if there are any schools from either dataframe missing these values:

In [2]:
df = df.merge(rank_df, how = 'left', left_index = True, right_index = True,
              validate = '1:1', indicator = True)

merged = (df['_merge'] == 'both')
df = df.drop(columns = '_merge')
rank_df.drop(df[merged].index, inplace = True)

len(rank_df)

299

Now we need to find a way to match these orphaned 299 US News ranks to the remaining CollegeData schools.

We can look to see if maybe the Name and State match but for some reason the City is off:

In [3]:
def strip(s):
    s = s.str.replace('SUNY','State University of New York')
    s = s.str.replace('CUNY','City University of New York')
    s = s.str.replace('\'s','s')
    s = s.str.replace('College|University|of|\'','')
    s = s.str.replace('St\.','Saint')
    s = s.str.replace('-|\.',' ')
    s = s.str.replace('\s+',' ')
    s = s.str.strip()
    return s

df['Stripped Name'] = strip(df.index.get_level_values('Name'))

match = lambda x: process.extract(x, df.loc[~merged, 'Stripped Name'])[0][0]
rank_df['Stripped Name'] = strip(rank_df.index.get_level_values('Name'))
rank_df['Stripped Name'] = rank_df['Stripped Name'].apply(match)

cols = ['Stripped Name', 'State', 'City']

df = df.reset_index()
dups = df.duplicated(cols, keep = False)
df.loc[dups, 'Stripped Name'] = df.loc[dups, 'Name']
df = df.set_index(cols)

rank_df = rank_df.reset_index()
dups = rank_df.duplicated(cols, keep = False)
rank_df.loc[dups, 'Stripped Name'] = rank_df.loc[dups, 'Name']
rank_df = rank_df.set_index(cols)

df = df.merge(rank_df.drop(columns = 'Name'), how = 'left', left_index = True,
              right_index = True, validate = '1:1', indicator = True)

merged = (df['_merge'] == 'both')
df = df.drop(columns = '_merge')
rank_df.drop(df[merged].index, inplace = True)

len(rank_df)

72

We can apply the same method to the City values:

In [4]:
df['Stripped City'] = strip(df.index.get_level_values('City'))

match = lambda x: process.extract(x, df.loc[~merged, 'Stripped City'])[0][0]
rank_df['Stripped City'] = strip(rank_df.index.get_level_values('City'))
rank_df['Stripped City'] = rank_df['Stripped City'].apply(match)

cols = ['Name', 'State', 'Stripped City']

df = df.reset_index()
dups = df.duplicated(cols, keep = False)
df.loc[dups, 'Stripped City'] = df.loc[dups, 'City']
df = df.set_index(cols)

rank_df = rank_df.reset_index()
dups = rank_df.duplicated(cols, keep = False)
rank_df.loc[dups, 'Stripped City'] = rank_df.loc[dups, 'City']
rank_df = rank_df.set_index(cols)

df = df.merge(rank_df.drop(columns = 'City'), how = 'left', left_index = True,
              right_index = True, validate = '1:1', indicator = True)

merged = (df['_merge'] == 'both')
df = df.drop(columns = '_merge')
rank_df.drop(df[merged].index, inplace = True)

len(rank_df)

63

In [5]:
df.head()

Stripped Name_x  \
Name                                           State Stripped City                                        
Bryn Athyn College                             PA    Bryn Athyn                              Bryn Athyn   
Adelphi University                             NY    Garden City                                Adelphi   
Albany College of Pharmacy and Health Sciences NY    Albany         Albany Pharmacy and Health Sciences   
Albertus Magnus College                        CT    New Haven                          Albertus Magnus   
Albright College                               PA    Reading                                   Albright   

                                                                           City  \
Name                                           State Stripped City                
Bryn Athyn College                             PA    Bryn Athyn      Bryn Athyn   
Adelphi University                             NY    Garden City    Garden City   
Albany College of Pharmacy and Health Sciences NY    Albany              Albany   
Albertus Magnus College                        CT    New Haven        New Haven   
Albright College                               PA    Reading            Reading   

                                                                    SchoolId  \
Name                                           State Stripped City             
Bryn Athyn College                             PA    Bryn Athyn            6   
Adelphi University                             NY    Garden City           7   
Albany College of Pharmacy and Health Sciences NY    Albany                8   
Albertus Magnus College                        CT    New Haven             9   
Albright College                               PA    Reading              10   

                                                                    2016 Graduates Who Took Out Loans  \
Name                                           State Stripped City                                      
Bryn Athyn College                             PA    Bryn Athyn                                  0.36   
Adelphi University                             NY    Garden City                                 0.61   
Albany College of Pharmacy and Health Sciences NY    Albany                                       NaN   
Albertus Magnus College                        CT    New Haven                                   0.83   
Albright College                               PA    Reading                                     0.88   

                                                                   24-Hour Emergency Phone/ Alarm Devices  \
Name                                           State Stripped City                                          
Bryn Athyn College                             PA    Bryn Athyn                                 Available   
Adelphi University                             NY    Garden City                                Available   
Albany College of Pharmacy and Health Sciences NY    Albany                                     Available   
Albertus Magnus College                        CT    New Haven                                  Available   
Albright College                               PA    Reading                                    Available   

                                                                   24-Hour Security Patrols  \
Name                                           State Stripped City                            
Bryn Athyn College                             PA    Bryn Athyn                         NaN   
Adelphi University                             NY    Garden City                  Available   
Albany College of Pharmacy and Health Sciences NY    Albany                       Available   
Albertus Magnus College                        CT    New Haven                    Available   
Albright College                               PA    Reading                      Available   

                             

In [5]:
collegedata_df.reset_index('City', inplace = True)
usnews_df.reset_index('City', inplace = True)
diff_cities = usnews_df.index.isin(collegedata_df.index)
diff_cities.sum()

12

Let's see the actual City values for these 12 cases to make sure that nothing is terribly off:

In [6]:
s1 = usnews_df.loc[diff_cities, 'City']
s2 = collegedata_df.loc[s1.index, 'City']
s1.name, s2.name = 'US News City', 'CollegeData City'
pd.concat([s2, s1], axis = 1)

CollegeData City  \
Name                            State                            
Florida Memorial University     FL                  Miami-Dade   
Saint Mary-of-the-Woods College IN     Saint Mary of the Woods   
University of Richmond          VA      University of Richmond   
Lake Superior State University  MI          Sault Sainte Marie   
Bard College                    NY         Annandale-on-Hudson   
St. Mary's College of Maryland  MD             St. Mary's City   
Auburn University               AL           Auburn University   
College of Mount St. Vincent    NY                   Riverdale   
Nova Southeastern University    FL             Fort Lauderdale   
American Jewish University      CA                     Bel Air   
Brewton-Parker College          GA                  Mt. Vernon   
Northern Illinois University    IL                     De Kalb   

                                                US News City  
Name                            State                         
Florida Memorial University     FL                     Miami  
Saint Mary-of-the-Woods College IN     St. Mary-of-the-Woods  
University of Richmond          VA         Univ. of Richmond  
Lake Superior State University  MI          Sault Ste. Marie  
Bard College                    NY       Annandale on Hudson  
St. Mary's College of Maryland  MD            St. Marys City  
Auburn University               AL                    Auburn  
College of Mount St. Vincent    NY                     Bronx  
Nova Southeastern University    FL            Ft. Lauderdale  
American Jewish University      CA                   Bel-Air  
Brewton-Parker College          GA              Mount Vernon  
Northern Illinois University    IL                    DeKalb

It looks like these are indeed the same schools - the City values are not matching because they have slightly different formatting. Let's join them into 'df' and remove them from the original dataframes:

In [7]:
joined = collegedata_df.join(usnews_df.drop(columns = 'City'), how = 'inner')
df = df.append(joined.set_index('City', append = True), sort = False)

collegedata_df.drop(joined.index, inplace = True)
usnews_df.drop(joined.index, inplace = True)

collegedata_df.reset_index(inplace = True)
usnews_df.reset_index(inplace = True)

has_ranks = df['Rank'].notna()
print("{} non-null Rank values in df.".format(has_ranks.sum()))
print("{} rows in collegedata_df, {} rows in collegedata_missing_df,\n"
      "{} rows in usnews_df, {} rows in df.".format(\
            len(collegedata_df), len(collegedata_missing_df),\
            len(usnews_df), len(df)))

1094 non-null Rank values in df.
880 rows in collegedata_df, 54 rows in collegedata_missing_df,
287 rows in usnews_df, 1986 rows in df.


We still have 287 ranks without a home. There's a good chance that the Name values in usnews_df are formatted slightly differently from the Name values in collegedata_df.

Using [fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy), which was put together by the people at SeatGeek and built on difflib, we can find a list of potential Name value matches in collegedata_df for each Name value in usnews_df:

In [8]:
def strip(s):
    s = s.str.replace('SUNY','State University of New York')
    s = s.str.replace('CUNY','City University of New York')
    s = s.str.replace('\'s','s')
    s = s.str.replace('College|University|of|\'','')
    s = s.str.replace('St\.','Saint')
    s = s.str.replace('-|\.',' ')
    s = s.str.replace('\s+',' ')
    s = s.str.strip()
    return s

collegedata_df['Stripped Name'] = strip(collegedata_df['Name'])

match = lambda x: process.extract(x, collegedata_df['Stripped Name'])[0][0]
usnews_df['Stripped Name'] = strip(usnews_df['Name']).apply(match)

Now we have a Stripped Name column in collegedata_df that contains the true stripped name. We also stripped the Names in usnews_df and immediately applied fuzzywuzzy's process.extract to find the closest matching Stripped Name in collegedata_df, and saved that possible match as usnews_df's Stripped Name column.

Now we'll set the indexes for both dataframes to include the Stripped Name column and then attempt an inner join (while dropping the useless original usnews_df Name column):

In [9]:
collegedata_df.set_index(['Stripped Name', 'State', 'City'], inplace = True)
usnews_df.set_index(['Stripped Name', 'State', 'City'], inplace = True)

joined = collegedata_df.join(usnews_df, how = 'inner', 
                             lsuffix = ' CollegeData', rsuffix = ' US News')

In [10]:
joined.reset_index(inplace = True)

In [13]:
joined[['Name CollegeData', 'Name US News', 'State', 'City']].duplicated().sum()

0

In [9]:
collegedata_df.drop(joined.index, inplace = True)
usnews_df.drop(joined.index, inplace = True)

Now we'll reset the joined index, drop the now useless Stripped Name column, set the normal Name, State, City index, and append joined to our combined 'df' dataframe:

In [10]:
joined.reset_index(inplace = True)
joined.drop(columns = 'Stripped Name', inplace = True)
joined.set_index(['Name', 'State', 'City'], inplace = True)
df = df.append(joined, sort = False)

collegedata_df.reset_index(inplace = True)
collegedata_df.drop(columns = 'Stripped Name', inplace = True)

usnews_df.reset_index(inplace = True)
usnews_df.drop(columns = 'Stripped Name', inplace = True)

has_ranks = df['Rank'].notna()
print("{} non-null Rank values in df.".format(has_ranks.sum()))
print("{} rows in collegedata_df, {} rows in collegedata_missing_df,\n"
      "{} rows in usnews_df, {} rows in df.".format(\
            len(collegedata_df), len(collegedata_missing_df),\
            len(usnews_df), len(df)))

1324 non-null Rank values in df.
651 rows in collegedata_df, 54 rows in collegedata_missing_df,
57 rows in usnews_df, 2216 rows in df.


That method seemed to help a bit - it cut our orphaned usnews_df rows from 287 to 57. We're approaching a threshold amount of values where it might be more efficient to manually rename them:

In [11]:
usnews_df.set_index('Name', inplace = True)
usnews_df.rename(index = school_rename_dict, inplace = True)
usnews_df = usnews_df[usnews_df.index.notna()]

usnews_df.set_index(['State'], append = True, inplace = True)
collegedata_df.set_index(['Name', 'State'], inplace = True)
joined = collegedata_df.join(usnews_df.drop(columns = 'City'), how = 'inner')

df = df.append(joined.set_index('City', append = True), sort = False)
collegedata_df.drop(joined.index, inplace = True)
usnews_df.drop(joined.index, inplace = True)

has_ranks = df['Rank'].notna()
print("{} non-null Rank values in df.".format(has_ranks.sum()))
print("{} rows in collegedata_df, {} rows in collegedata_missing_df,\n"
      "{} rows in usnews_df, {} rows in df.".format(\
            len(collegedata_df), len(collegedata_missing_df),\
            len(usnews_df), len(df)))

1351 non-null Rank values in df.
624 rows in collegedata_df, 54 rows in collegedata_missing_df,
7 rows in usnews_df, 1351 rows in df.


We can try to join the remaining 7 rows from usnews_df on the rows we put into collegedata_missing_df that were missing City and/or State values:

In [12]:
usnews_df.reset_index(inplace = True)
usnews_df.set_index('Name', inplace = True)
collegedata_missing_df.set_index('Name', inplace = True)

joined = usnews_df.join(collegedata_missing_df\
                        .drop(columns = ['State', 'City']), how = 'inner')

df = df.append(joined\
               .set_index(['State', 'City'], append = True), sort = False)

usnews_df.drop(joined.index, inplace = True)

usnews_df.reset_index(inplace = True)

has_ranks = df['Rank'].notna()
print("{} non-null Rank values in df.".format(has_ranks.sum()))
print("{} rows in collegedata_df, {} rows in collegedata_missing_df,\n"
      "{} rows in usnews_df, {} rows in df.".format(\
            len(collegedata_df), len(collegedata_missing_df),\
            len(usnews_df), len(df)))

1356 non-null Rank values in df.
624 rows in collegedata_df, 54 rows in collegedata_missing_df,
2 rows in usnews_df, 1356 rows in df.


There are only two schools remaining in usnews_df:

In [13]:
usnews_df

Name State    City  Rank                      Rank Type
0  Concordia University    NE  Seward    38  Regional Universities Midwest
1  Concordia University    CA  Irvine    41     Regional Universities West

These two share the same name in usnews_df, but their corresponding rows in collegedata_df are different. We'll change the names, join them to 'df':

In [14]:
usnews_df.loc[0, 'Name'] = 'Concordia University Nebraska'
usnews_df.loc[1, 'Name'] = 'Concordia University Irvine'

usnews_df.set_index(['Name', 'State', 'City'], inplace = True)
collegedata_df.set_index('City', append = True, inplace = True)

joined = collegedata_df.join(usnews_df, how = 'inner')

df = df.append(joined, sort = False)

We'll also want to add our remaining collegedata_missing_df rows back to df, even 

In [15]:
len(df)



1358

In [19]:
df[['Name','State','City']].duplicated().sum()

271

In [20]:
len(collegedata_missing_df)

54

In [13]:
A = [8, 2, 2]

def test(x):
    
    print(len(A))
    
    return None



A = [3]

test(42)


1
